shared/js/background/storage/cookies.es6.js

```diff
@@ -0,0 +1,81 @@
+// Babel includes for async/await
+import 'regenerator-runtime/runtime'
+
+const load = require('./../load.es6')
+const constants = require('../../../data/constants')
+const settings = require('./../settings.es6')
+
+/**
+ *  Manage local cookie exclusion data.
+ *  Data is updated regularly with the latest
+ *  cookie exclusions, and stored here for fast retrieval and usage
+ *  when checking which cookies to block.
+ **/
+class ExcludedCookieStorage {
+    constructor () {
+        this.excludedDomains = []
+    }
+
+    /**
+     * Retrieve the latest lists of excluded cookie domains. Store as needed.
+     */
+    updateCookieData () {
+        console.log('Cookies: Getting cookie data')
+        const lists = constants.CookieLists
+        for (const list of lists) {
+            const source = list.source || 'external'
+            const listName = list.name
+            const etag = settings.getSetting(`${listName}-etag`) || ''
+            load.loadExtensionFile({ url: list.url, etag: etag, returnType: list.format, source, timeout: 60000 })
+                .then(response => {
+                    if (response && (response.status === 200 || response.status === 304)) {
+                        // New cookie data to process.
+                        const data = JSON.parse(response.response)
+                        this.processList(listName, data)
+                        const newEtag = response.getResponseHeader('etag') || ''
+                        settings.updateSetting(`${listName}-etag`, newEtag)
+                    }
+                })
+                .catch(e => {
+                    // Reset the etag
+                    settings.updateSetting(`${listName}-etag`, '')
+                    console.log(`Error updating cookie data:  ${e}. Attempting to load from local storage.`)
+                })
+        }
+    }
+
+    processList (listName, listData) {
+        switch (listName) {
+        case 'cookieExcludeList':
+            this.processExcludeList(listData)
+            break
+        }
+    }
+
+    async processExcludeList (data) {
+        this.excludedDomains = []
+        for (const record of data.excludedDomains) {
+            this.excludedDomains.push(record.domain)
+        }
+    }
+
+    isExcluded (url) {
+        const domain = (new URL(url)).host
+        return this.isDomainExcluded(domain)
+    }
+
+    isDomainExcluded (domain) {
+        if (this.excludedDomains.find(elem => elem === domain)) {
+            return true
+        }
+
+        const comps = domain.split('.')
+        if (comps.length > 2) {
+            comps.shift()
+            return this.isDomainExcluded(comps.join('.'))
+        }
```

Comment:
This function is O(N*M) in the worst case, where N is the length of `excludedDomains` and M is the subdomain depth of the given domain. As this is run for every third-party request, then most calls will run to the worst case complexity, and this function will be called often. 

I think it would make sense to optimise this a bit. Using a trie you should be able to get O(log M) performance on average.

Reply:
I'm having issues imagining the trie implementation (my theory is rather rusty). Would you be able to go a bit more in depth. Do you think we'll be hitting the worst case most often because most cookies will be excluded by the root domain (i.e. example.com vs test.example.com)? If so the other idea I have is recursing to the root domain first and bubbling up from there. However I think some granularity in the exclusions could be lost this way?

Reply:
I would organise the exclude list by reversed domain parts:
```json
{
  "com": {
     "google": {
        "accounts": true
     },
     "googleapis": true
}
```

Then you get worst-case O(M), but constant time average case with a lookup algorithm like:
```js
const parts = domain.split('.').reverse()
let ptr = this.excludedDomainsLookup;
for (const i = 0; i < parts.length; i++) {
  ptr = ptr[parts[i]]
  if (ptr === true) { // matching subdomain
     return true
  } else if (ptr === undefined) { // empty leaf
     return false;
  }
}
```
(warning: untested code that I just dropped in here, might not work at all!)